src/shred_network/duplicate_shred_listener.zig

```diff
@@ -0,0 +1,794 @@
+    try std.testing.expectEqual(null, handler.dup_buffer.get(key));
+}
+
+test "DuplicateShredHandler: cacheRootInfo updates cached slots in epoch" {
+    const allocator = std.testing.allocator;
+    var epoch_ctx = try sig.adapter.EpochContextManager.init(allocator, .DEFAULT);
+    defer epoch_ctx.deinit();
+
+    var ledger = try sig.ledger.tests.initTestLedger(allocator, @src(), .noop);
+    defer ledger.deinit();
+
+    var dup_slots_channel = try Channel(Slot).init(allocator);
+    defer dup_slots_channel.deinit();
+
+    var shred_version = std.atomic.Value(u16).init(0);
+
+    var handler = DuplicateShredHandler.init(allocator, .noop, .{
+        .exit = undefined,
+        .gossip_table_rw = undefined,
+        .result_writer = ledger.resultWriter(),
+        .ledger_reader = ledger.reader(),
+        .duplicate_slots_sender = &dup_slots_channel,
+        .leader_schedule = epoch_ctx.slotLeaders(),
+        .shred_version = &shred_version,
+        .epoch_schedule = epoch_ctx.schedule,
+        .epoch_ctx_mgr = &epoch_ctx,
+    });
+    defer handler.deinit();
+
+    {
+        var setter = try handler.params.result_writer.setRootsIncremental();
+        defer setter.deinit();
+        try setter.addRoot(1);
+        try setter.commit();
+
+        handler.cacheRootInfo();
+        const expected_epoch = handler.params.epoch_schedule.getEpoch(1);
+        const expected_slots_in_epoch =
+            handler.params.epoch_schedule.getSlotsInEpoch(expected_epoch);
+        try std.testing.expectEqual(@as(Slot, 1), handler.last_root);
+        try std.testing.expectEqual(expected_slots_in_epoch, handler.cached_slots_in_epoch);
+    }
+
+    {
+        const update_epoch: sig.core.Epoch = 2;
+        const update_slot = handler.params.epoch_schedule.getFirstSlotInEpoch(update_epoch);
+        var setter2 = try handler.params.result_writer.setRootsIncremental();
+        defer setter2.deinit();
+        try setter2.addRoot(update_slot);
+        try setter2.commit();
+
+        handler.cacheRootInfo();
+        try std.testing.expectEqual(update_slot, handler.last_root);
+        const expected_slots_in_epoch2 = handler.params.epoch_schedule.getSlotsInEpoch(update_epoch);
+        try std.testing.expectEqual(expected_slots_in_epoch2, handler.cached_slots_in_epoch);
+    }
+}
+
+test "DuplicateShredHandler: maybePruneBuffer prunes when over capacity" {
+    const allocator = std.testing.allocator;
+    var epoch_ctx = try sig.adapter.EpochContextManager.init(allocator, .DEFAULT);
+    defer epoch_ctx.deinit();
+
+    var ledger = try sig.ledger.tests.initTestLedger(allocator, @src(), .noop);
+    defer ledger.deinit();
+
+    var dup_slots_channel = try Channel(Slot).init(allocator);
+    defer dup_slots_channel.deinit();
+
+    var shred_version = std.atomic.Value(u16).init(0);
+
+    var handler = DuplicateShredHandler.init(allocator, .noop, .{
+        .exit = undefined,
+        .gossip_table_rw = undefined,
```

Comment:
Would it be possible to split out the relevant subset of functionality from `DuplicateShredHandler` into its own thing that can be easily tested without needing to initialize these fields?